### ISSUE TITLE
Bad performance on macOS

### ISSUE THREAD

**Issue by tarcieri** [miscreant/miscreant.rs#20]
Encryption benchmarks for **miscreant.rs** versions v0.3.0 and v0.4.2 on Linux and macOS (note Linux performance is not similarly degraded):

| Version | OS   | AES-CMAC-SIV CPB | AES-PMAC-SIV CPB |
|-------|-------|----------------------|----------------------|
| v0.3.0 | Linux | 2.7                               | 1.3                                |
| v0.4.2 | Linux | 3.0                               | 1.5                                |
| v0.3.0 | macOS | 14.8                          | 54.4                             |
| v0.4.2 | macOS | 28.3                          | 92.8                            |

*NOTE: lower is better for cycles-per-byte count below, and the slight performance degradation on Linux is expected due to some unrelated cipher initialization changes*

Benchmarks were performed on `rustc 1.33.0-nightly (09d6ab90e 2018-12-20)` for v0.3.0, and both that nightly and 1.31.0 for v0.4.2, with similar results (figures shown are nightly).

**Comment by tarcieri** [comment]
Note that I do not believe that this poor performance on macOS was always the case: I did a lot of benchmarking on macOS previously, including of v0.3.0, and it was previously in line with Linux and the figures above.

These benchmarks were taken on Mojave (10.14.2). Previously I was using an earlier version of macOS.

**Comment by newpavlov** [comment]
Can you also measure performance of `aes`, `cmac` and `pmac`crates on macOS?

**Comment by tarcieri** [comment]
Seems the existing benchmarks don't work on `stable`. Anywho, here they are on `nightly-2018-12-20`:

tl;dr: PMAC seems slow... the others are about what I'd expect...

## `aesni`

```
test aes128_decrypt  ... bench:          13 ns/iter (+/- 1) = 1230 MB/s
test aes128_decrypt8 ... bench:          24 ns/iter (+/- 2) = 5333 MB/s
test aes128_encrypt  ... bench:          13 ns/iter (+/- 1) = 1230 MB/s
test aes128_encrypt8 ... bench:          24 ns/iter (+/- 3) = 5333 MB/s
```

```
test aes256_decrypt  ... bench:          18 ns/iter (+/- 7) = 888 MB/s
test aes256_decrypt8 ... bench:          34 ns/iter (+/- 5) = 3764 MB/s
test aes256_encrypt  ... bench:          18 ns/iter (+/- 2) = 888 MB/s
test aes256_encrypt8 ... bench:          32 ns/iter (+/- 16) = 4000 MB/s
```

## `cmac`

### AES-128

```
test bench3_100   ... bench:         154 ns/iter (+/- 27) = 649 MB/s
test bench4_1000  ... bench:       1,530 ns/iter (+/- 181) = 653 MB/s
test bench5_10000 ... bench:      15,384 ns/iter (+/- 1,662) = 650 MB/s
```

### AES-256

```
test bench3_100   ... bench:         183 ns/iter (+/- 31) = 546 MB/s
test bench4_1000  ... bench:       1,835 ns/iter (+/- 214) = 544 MB/s
test bench5_10000 ... bench:      18,154 ns/iter (+/- 2,406) = 550 MB/s
```

## `pmac`

These seem abnormally low...

### AES-128

```
test bench3_100   ... bench:         185 ns/iter (+/- 29) = 540 MB/s
test bench4_1000  ... bench:       1,780 ns/iter (+/- 420) = 561 MB/s
test bench5_10000 ... bench:      17,819 ns/iter (+/- 2,071) = 561 MB/s
```

### AES-256

```
test bench3_100   ... bench:         194 ns/iter (+/- 27) = 515 MB/s
test bench4_1000  ... bench:       1,943 ns/iter (+/- 311) = 514 MB/s
test bench5_10000 ... bench:      19,916 ns/iter (+/- 5,053) = 502 MB/s
```

**Comment by newpavlov** [comment]
Hm, I thought `pmac` was faster than that, but after re-running benchmarks I get approximately the same numbers... Testing on nightlies from several months ago does not change result much, so we can rule-out drastic regression.

I have some ideas on how `pmac` can be improved, after looking at the code it's quite sub-optimal in several places, so I will try to implement improvements.

**Comment by tarcieri** [comment]
@newpavlov here's what's really weird. Here's Miscreant's AES-PMAC-SIV benchmarks (using `criterion`)

```
AES-128-PMAC-SIV/encrypt (1024 bytes)
                        time:   [970.80 ns 970.96 ns 971.16 ns]
                        thrpt:  [1005.6 MiB/s 1005.8 MiB/s 1005.9 MiB/s]
                        change: [+0.2395% +0.2608% +0.2942%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 15 outliers among 100 measurements (15.00%)
  1 (1.00%) low severe
  4 (4.00%) high mild
  10 (10.00%) high severe

AES-128-PMAC-SIV/encrypt (16384 bytes)
                        time:   [11.780 us 11.781 us 11.781 us]
                        thrpt:  [1.2952 GiB/s 1.2952 GiB/s 1.2953 GiB/s]
                        change: [-0.2255% -0.2082% -0.1785%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```

But here's the `pmac` crate benchmarks on the same compiler/CPU (nightly-2019-01-12 on Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz)

```
     Running /home/bascule/MACs/target/release/deps/aes128_pmac-180fc8444bd49974

running 4 tests
test bench1_10    ... bench:          17 ns/iter (+/- 0) = 588 MB/s
test bench2_100   ... bench:         144 ns/iter (+/- 0) = 694 MB/s
test bench3_1000  ... bench:       1,394 ns/iter (+/- 1) = 717 MB/s
test bench3_10000 ... bench:      13,745 ns/iter (+/- 8) = 727 MB/s

test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out

     Running /home/bascule/MACs/target/release/deps/aes256_pmac-6d7b86b8a8804bc9

running 4 tests
test bench1_10    ... bench:          18 ns/iter (+/- 0) = 555 MB/s
test bench2_100   ... bench:         151 ns/iter (+/- 0) = 662 MB/s
test bench3_1000  ... bench:       1,454 ns/iter (+/- 1) = 687 MB/s
test bench3_10000 ... bench:      14,335 ns/iter (+/- 5) = 697 MB/s
```

...so I'm not sure we're getting accurate benchmarks yet: that shows the full AES-PMAC-SIV running considerably faster than just PMAC alone.

If you're interested, I could submit a PR to `MACs` to switch the benchmarks to `criterion` and we could see what the results are for comparison. The extremely slow results in the OP are also under `criterion`, so perhaps we can see if that's the source of the extremely pathological slowness (which might indicate some sort of `criterion` bug).

Conversely, If it turns out that the `benches` feature is buggy, it might make sense to switch to `criterion`. It also has the advantage of being able to run on `stable` and it's maintained.

**Comment by newpavlov** [comment]
Note that benchmarks are not equivalent to each other, for MACs code looks like [this](https://github.com/RustCrypto/traits/blob/master/crypto-mac/src/dev.rs#L60). It measures performance of the `input` method for different positions of inner buffer without MAC finalization, this is why input buffer size is not power of the block size and MAC state gets reused across measurements, so this benchmark has higher difficulty. It's probably not the best benchmark approach, so we could improve it, but I am on the fence regarding using `criterion`. I think if we'll decide to use it, we should migrate all crates by changing `dev` module in trait crates.

**Comment by tarcieri** [comment]
All right, that at least explains the discrepancies, thanks. That said, I'm not particularly motivated to continue a benchmarking-based approach to breaking down this problem.

I think the next thing I'd like to try is running the problematic code in a continuous loop and profiling it with Instruments

**Comment by tarcieri** [comment]
This issue is now obsolete.
